src/tracks/InstrumentTrack.cpp

```diff
@@ -579,6 +579,13 @@ void InstrumentTrack::playNote( NotePlayHandle* n, sampleFrame* workingBuffer )
 	{
 		// all is done, so now lets play the note!
 		m_instrument->playNote( n, workingBuffer );
+
+		if (workingBuffer != nullptr)
```

Comment:
Hi @sakertooth, as was mentioned in the commit message (see d3d710e0ada) this was done to keep the old behavior because I do not know if the call to `playNote` with a `nullptr` still has desirable side effects as this is what also happens in the original code:

> In `InstrumentTrack::playNote` an additional check was added which only calls `processAudioBuffer` if the buffer is not `nullptr`. The reason is that under certain circumstances `PlayHandle::doProcessing` calls the `play` method by explicitly passing a `nullptr` as the buffer. This behavior was added with commit https://github.com/LMMS/lmms/commit/7bc97f5d5bd9fcb43971f62e546ba464be1e818f. Because it is unknown if this was done for some side effects the code was adjusted so that it behaves identical in this case.

I don't know why the `play` method is called with a `nullptr` in the first place.

I also forgot to mention in the commit message that if `processAudioBuffer` is called with a `nullptr` buffer then the helper method that checks for silence crashes.

So I'd like to keep the code like this to be on the safe side.

Reply:
>  The reason is that under certain circumstances PlayHandle::doProcessing calls the play method by explicitly passing a nullptr as the buffer.

Well, if anything, move that check to the parent if-statement because we are still calling ``Instrument::playNote``  with a buffer that's potentially ``nullptr``. In that instance, we would be crashing. These calls do not check against that, and you shouldn't rely on them doing so. Also, the current code processes empty buffers, even though the original code did not do that, as calls to ``processAudioBuffer`` where only made if the ``Instrument`` made a successful call to ``playNote`` and did not ``memset`` the entire buffer to 0 before exiting. This is why I suggested returning a ``bool`` from ``Instrument::playNote`` and checking if it returned ``true`` before calling ``processAudioBuffer``.

Reply:
Hi @sakertooth,

what I am trying to tell you is that `Instrument::playNote` is already called with a `nullptr` in master, i.e. even without my changes. It also does not crash while doing so. I don't know if there are any side effects while it is executed with the `nullptr` and therefore I don't want to touch this any more than necessary and only add this incremental improvement.

If you want to make more sweeping changes then lets merge this PR and then you can do them. Or even better: find out why `play` is called with a `nullptr` in `PlayHandle` since commit 7bc97f5d5bd in the first place. I think @diizy is not around anymore to tell us why the code was adjusted like this. The fact that it was not simple removed makes me assume that it has some side effects.

If you want to try this out you can do this as follows:
1. Check out commit 3a0e68c0ac9, i.e. the commit before my changes.
2. Add some code which prints something or hits a break point if the buffer is `nullptr`.
3. Build and run LMMS.
4. Open "Greippi - Krem Kaakkuja (Second Flight Remix).mmpz" from the demos.
5. Set the play head to before bar 9 and play the song.
6. Once bar 9 is hit you will see output or the break point being hit.

Reply:
> what I am trying to tell you is that Instrument::playNote is already called with a nullptr in master, i.e. even without my changes. It also does not crash while doing so. I don't know if there are any side effects while it is executed with the nullptr and therefore I don't want to touch this any more than necessary and only add this incremental improvement.

Okay, I understand now. You don't want to change any current behavior of ``Instrument::playNote`` being called with a ``nullptr``, but want to guard against it for ``processAudioBuffer``. You may want to add a comment to explain this further.

However, can the processing of empty buffers be addressed? Functions like ``AudioFileProcessor::playNote`` could end up making more unnecessary calls to ``InstrumentTrack::processAudioBuffer`` because of this than before.

Reply:
I have added some comments with commit 6c3ae30c893.

Addressing the processing of `nullptr` buffers might be a good idea. But it is not in the scope of this PR which simply brought awareness to this preexisting problem.

Reply:
> Addressing the processing of nullptr buffers might be a good idea. But it is not in the scope of this PR which simply brought awareness to this preexisting problem.

No, I mean when calls to ``Instrument::playNote`` ``memset`` the entire buffer to be empty, they do not go ahead and call ``processAudioBuffer``. However, here, we call it no matter what. This does not keep the same behavior, and we shouldn't be calling ``processAudioBuffer`` even when we don't have to.

Reply:
If I understand correctly `processAudioBuffer` does the following:
1. Check if the buffer consists of silence and stop if that is the case.
2. Apply the sound shaping, i.e. filters and envelopes.
3. Apply volume and panning.

If that is correct then not calling `processAudioBuffer`, e.g. because the buffer is filled with zeros, is the wrong thing to do because the filter cannot finish its job to turn harsh steps into softer sounds.

Reply:
> If that is correct then not calling processAudioBuffer, e.g. because the buffer is filled with zeros, is the wrong thing to do because the filter cannot finish its job to turn harsh steps into softer sounds.

For "turning harsh steps into softer sounds", are you referring to the transition from a non-empty buffer to an empty one so there's no click in the audio? In that case, I suppose you're right. If not, then I'm confused.

The ``nullptr`` situation is a bit confusing too. I did the steps you mentioned to reproduce it, and for some reason, it isn't crashing when calling ``Instrument::playNote`` with a buffer that is a ``nullptr``. Not sure why extracting a call to ``processAudioBuffer`` would cause a crash though, unless ``playNote`` was protecting against a ``nullptr`` for a buffer somehow. It can be looked at later I guess.

Reply:
Yes, I am referring to the transition between non-zeroed and zeroed buffers. Let's assume there's an instrument that fills a first buffer completely with non-zero samples and then zeros the second buffer completely. In this case there's a harsh transition between the last (non-zero) sample of the first buffer and the first (zero) sample in the second buffer. This harsh transition is "smeared out" by a filter. However, to be able to do so the filter must have a chance to work on the second buffer so that the first samples are replaced with non-zero samples.

With regards to the `nullptr` situation my understanding is the following. There seem to be two interfaces/methods that instruments can use to implement their sound generation:
* `Instrument::playNote`
* `Instrument::play`

The ones that use `playNote` are:
* DummyInstrument
* AudioFileProcessor
* BitInvader
* FreeBoy
* GigPlayer
* Kicker
* Lb302
* Monstro
* Nes
* Organic
* Patman
* Sf2Player
* Sfxr
* SidInstrument
* Mallets
* TripleOscillator
* Vibed
* Watsyn
* Xpressive

The ones that use `play` are:
* Carla
* GigPlayer
* Lb302
* Lv2Instrument
* OpulenZ
* Sf2Player
* Vestige
* ZynAddSubFx

As you can see there are some instruments that support both interfaces:
* GigPlayer
* Lb302
* Sf2Player

I think these three instruments are the ones that lead to the usage of buffers with `nullptr` in the first place. If you check their implementation of `playNote` you will find that they do not access the buffer at all and hence there are no crashes in this scenario.

Reply:
Okay, I discovered that within the Greippi demo, ``Lb302`` is what triggers a call to ``play(nullptr)`` because it is an ``InstrumentPlayHandle``, and ``NotePlayHandle`` recognizes that and sets ``m_usesBuffer`` to false in this way (on ``NotePlayHandle.cpp``, L#112), causing a ``nullptr`` to be passed once ``PlayHandle::doProcessing`` is called.

```cpp
if(m_instrumentTrack->instrument() && m_instrumentTrack->instrument()->flags() & Instrument::Flag::IsSingleStreamed )
{
  setUsesBuffer( false );
}
```
So you are right. ``Instrument::playNote`` should still be called even when the buffer is ``nullptr`` (or else we would break the processing of ``InstrumentPlayHandle``'s). The only difference is that for ``InstrumentPlayHandle``, ``processAudioBuffer`` is not called, so you calling it without the ``workingBuffer != nullptr`` guard was most likely why it crashed.

I would suggest using the same condition shown above instead of ``workingBuffer != nullptr``, i.e., only call ``processAudioBuffer`` if it is not an ``InstrumentPlayHandle``. This seems to be the intended behavior, and it is more descriptive that way.

Reply:
Hi @sakertooth,

commit b6c80bd7366 replaces the check with something more fitting. Please refer to the commit message for the rationale.